text/0032-dust-protection/0032-dust-protection.md

```diff
@@ -0,0 +1,179 @@
+            if numOfDustDeposits > numOfAllowedDustOutputs:
+                return false
+        
+     return true              
+```
+
+# Drawbacks
+
+- There can no longer be addresses holding less than 1 Mi.
+- A service receiving micropayments may fail receiving them, if it did not consolidate dust outputs or raised the deposit for the receiving address.
+
+# Rationale and alternatives
+
+The rationale for creating a special `SigLockedDustAllowanceOutput` rather than rely on the default `SigLockedSingleOutputs` is to prevent attackers from polluting arbitrary addresses that happen to hold
+a large amount of funds with dust.
+
+One may note that an attacker can deposit a dust allowance on 3rd party address outside his control and pollute that address with dust.
+From a security perspective this is better than an attacker depositing a dust allowance on addresses under his control.
+This is because the receiving party might later choose to consolidate the dust outputs and hence relief UTXO memory consumption.
+The receiving party is also unlikely to be displeased from obtaining more funds, small as they may be.
+
+There are potential alternatives to introducing dust allowance deposits:
+
+- *Burning dust*: Allow dust outputs to exists only for a limited amount of time in the ledger. After this, they are removed completely and the associated funds are invalidated.
+- *Sweeping dust into Merkle trees*: Instead of burning dust outputs after some time, they are instead compressed into a Merkle tree and only the tree root is kept. In order to spend one of these compressed outputs, the corresponding Merkle audit path needs to be supplied in addition to a regular signature.
+
+The first option can cause issues, when dust outputs were burned before users could consolidate them. Also, changing the supply can be controversial.
+
+The second option is much more complicated as it introduces a completely new unlock mechanisms and requires the nodes to store the Merkle tree roots indefinitely.
+
+
+# Unresolved questions
+
+- An attacker can send microtransactions to an address with a `SigLockedDustAllowanceOutput` in order to fill the allowed threshold and block honest senders of microtransactions. The owner of the address can mitigate this by simply consolidating the attacker's dust and collecting it for profit. The problem is that the cost of doing PoW too often may exceed the profit made by collecting dust. Perhaps it is a good idea to limit "tiny" (1 i) dust outputs more than larger dust outputs? In the original discussion each dust class was supposed to be capped to a certain fixed amount. But perhaps we should give weights to different dust classes. So "tiny" dust will weigh more and fill the cap more quickly than larger dust?
+
+- Total cap per address so db key won't be polluted... should this be part of the RFC? This also depends on how key-value dbs will behave at the implementation level. RocksDb for example can configured to handle this well.
```

Comment:
Not part of the RFC...

We might need to paginate the db keys... we should test on implementations.